pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkloadMultiRegion failed

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkloadMultiRegion [failed](https://mesolite.cluster.engflow.com/invocations/default/0c449750-f7f7-49c3-910f-2fc419192201?testReportRun=1&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvd29ya2xvYWQvc2NoZW1hY2hhbmdlOnNjaGVtYWNoYW5nZV90ZXN0) on master @ [79c8e71012ccdba87e6754f36995e0da0de63e3f](https://github.com/cockroachdb/cockroach/commits/79c8e71012ccdba87e6754f36995e0da0de63e3f):


```
 }
}
    schema_change_external_test.go:183: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:183
        	            				pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:50
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED ERROR; Received an unexpected execution error.: ERROR: internal error: executing declarative schema change StatementPhase stage 1 of 1 with 11 MutationType ops (rollback=false) for CREATE TRIGGER: error executing StatementPhase stage 1 of 1 with 11 MutationType ops: relation "table_w0_45" (121): invalid locality config: expected no region Enum ID to be referenced by a REGIONAL BY TABLE: "table_w0_45" homed in the primary region, but found: 106 (SQLSTATE XX000)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*opStmt).executeStmt
        	            	  | 	pkg/workload/schemachange/operation_generator.go:3543
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	pkg/workload/schemachange/schemachange.go:612
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:700
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.runSchemaChangeWorkload.runSchemaChangeWorkload.func5.func7
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:172
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1771
        	            	Wraps: (4) ***UNEXPECTED ERROR; Received an unexpected execution error.
        	            	Wraps: (5) ERROR: internal error: executing declarative schema change StatementPhase stage 1 of 1 with 11 MutationType ops (rollback=false) for CREATE TRIGGER: error executing StatementPhase stage 1 of 1 with 11 MutationType ops: relation "table_w0_45" (121): invalid locality config: expected no region Enum ID to be referenced by a REGIONAL BY TABLE: "table_w0_45" homed in the primary region, but found: 106 (SQLSTATE XX000)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkloadMultiRegion
{
 "message": "Inspect Job Errors",
 "jobs": [
  {
   "jobId": 1167530450286804993,
   "status": "no errors reported"
  }
 ]
}
Schema Workload Stats
Total Schema Statements Executed = 31
Total Schema Statements Succeeded = 31
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 21
Total Transactions Rolled Back = 22
Total Transactions Executed = 43
    schema_change_external_test.go:136: backup, tracing data, and system table dumps in /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/ccl/testccl/workload/schemachange/schemachange_test/shard_2_of_2/test.outputs/logTestWorkloadMultiRegion1171663758
    schema_change_external_test.go:145: running DROP with use_declarative_schema_changer = off
    panic.go:694: -- test log scope end --
test logs left over in: outputs.zip/logTestWorkloadMultiRegion1171663758
--- FAIL: TestWorkloadMultiRegion (17.53s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkloadMultiRegion.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-62993

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkloadMultiRegion [failed](https://mesolite.cluster.engflow.com/invocations/default/5ad489e7-c657-4cbb-bef8-1cb59fe0ebd4?testReportRun=1&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvd29ya2xvYWQvc2NoZW1hY2hhbmdlOnNjaGVtYWNoYW5nZV90ZXN0) on master @ [8b6d3d2a14f67c0867d9f492fde1fad929538a51](https://github.com/cockroachdb/cockroach/commits/8b6d3d2a14f67c0867d9f492fde1fad929538a51):


```
   },
   {
    "message": "createTrigger: CREATE TRIGGER trigger_w1_250 AFTER INSERT OR UPDATE ON public.table_w0_148 FOR EACH ROW EXECUTE FUNCTION public.trigger_function_w0_126()"
   }
  ],
  "previousStatements": [
   "CREATE TRIGGER trigger_w1_250 AFTER INSERT OR UPDATE ON public.table_w0_148 FOR EACH ROW EXECUTE FUNCTION public.trigger_function_w0_126()"
  ],
  "usesDeclarativeSchemaChanger": true
 }
}
    schema_change_external_test.go:183: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:183
        	            				pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:50
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED ERROR; Received an unexpected execution error.: ERROR: internal error: executing declarative schema change StatementPhase stage 1 of 1 with 11 MutationType ops (rollback=false) for CREATE TRIGGER: error executing StatementPhase stage 1 of 1 with 11 MutationType ops: relation "table_w0_148" (136): invalid locality config: expected no region Enum ID to be referenced by a REGIONAL BY TABLE: "table_w0_148" homed in the primary region, but found: 106 (SQLSTATE XX000)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*opStmt).executeStmt
        	            	  | 	pkg/workload/schemachange/operation_generator.go:3543
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	pkg/workload/schemachange/schemachange.go:612
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:700
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.runSchemaChangeWorkload.runSchemaChangeWorkload.func5.func7
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:172
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1771
        	            	Wraps: (4) ***UNEXPECTED ERROR; Received an unexpected execution error.
        	            	Wraps: (5) ERROR: internal error: executing declarative schema change StatementPhase stage 1 of 1 with 11 MutationType ops (rollback=false) for CREATE TRIGGER: error executing StatementPhase stage 1 of 1 with 11 MutationType ops: relation "table_w0_148" (136): invalid locality config: expected no region Enum ID to be referenced by a REGIONAL BY TABLE: "table_w0_148" homed in the primary region, but found: 106 (SQLSTATE XX000)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkloadMultiRegion
Schema Workload Stats
Total Schema Statements Executed = 83
Total Schema Statements Succeeded = 83
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 50
Total Transactions Rolled Back = 46
Total Transactions Executed = 96
    schema_change_external_test.go:136: backup, tracing data, and system table dumps in /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/ccl/testccl/workload/schemachange/schemachange_test/shard_2_of_2_run_1_of_10/test.outputs/logTestWorkloadMultiRegion110786934
    schema_change_external_test.go:145: running DROP with use_declarative_schema_changer = off
    panic.go:694: -- test log scope end --
test logs left over in: outputs.zip/logTestWorkloadMultiRegion110786934
--- FAIL: TestWorkloadMultiRegion (45.48s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkloadMultiRegion.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkloadMultiRegion [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21314772?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21314772?buildTab=artifacts#/) on master @ [8b6d3d2a14f67c0867d9f492fde1fad929538a51](https://github.com/cockroachdb/cockroach/commits/8b6d3d2a14f67c0867d9f492fde1fad929538a51):


```
 }
}
    schema_change_external_test.go:183: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:183
        	            				pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:50
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED COMMIT ERROR; Received an unexpected commit error: ERROR: transaction committed but schema change aborted with error: (2BP01): could not remove enum value "k" as it is being used in a routine "udf_w1_98" (SQLSTATE 2BP01)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:771
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.runSchemaChangeWorkload.runSchemaChangeWorkload.func5.func7
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:172
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_s390x.s:864
        	            	Wraps: (4) ***UNEXPECTED COMMIT ERROR; Received an unexpected commit error
        	            	Wraps: (5) ERROR: transaction committed but schema change aborted with error: (2BP01): could not remove enum value "k" as it is being used in a routine "udf_w1_98" (SQLSTATE 2BP01)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkloadMultiRegion
{
 "message": "Inspect Job Errors",
 "jobs": [
  {
   "jobId": 1167633733493030913,
   "status": "no errors reported"
  },
  {
   "jobId": 1167633739378098177,
   "status": "no errors reported"
  }
 ]
}
Schema Workload Stats
Total Schema Statements Executed = 99
Total Schema Statements Succeeded = 99
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 62
Total Transactions Rolled Back = 74
Total Transactions Executed = 136
    schema_change_external_test.go:136: backup, tracing data, and system table dumps in /artifacts/tmp/_tmp/bf9e40dc5ebd9e5f3b9b38e49138277f/logTestWorkloadMultiRegion1894682060
    schema_change_external_test.go:145: running DROP with use_declarative_schema_changer = off
    panic.go:694: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/bf9e40dc5ebd9e5f3b9b38e49138277f/logTestWorkloadMultiRegion1894682060
--- FAIL: TestWorkloadMultiRegion (109.24s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkloadMultiRegion.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkloadMultiRegion [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21317595?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21317595?buildTab=artifacts#/) on master @ [980e96a3482d2fed84c62da15ca95c0ee40c0639](https://github.com/cockroachdb/cockroach/commits/980e96a3482d2fed84c62da15ca95c0ee40c0639):


```
 }
}
    schema_change_external_test.go:183: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:183
        	            				pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:50
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED COMMIT ERROR; Received an unexpected commit error: ERROR: transaction committed but schema change aborted with error: (2BP01): could not remove enum value "chpw" as it is being used in a routine "udf_w1_328" (SQLSTATE 2BP01)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:771
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.runSchemaChangeWorkload.runSchemaChangeWorkload.func5.func7
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:172
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_arm64.s:1447
        	            	Wraps: (4) ***UNEXPECTED COMMIT ERROR; Received an unexpected commit error
        	            	Wraps: (5) ERROR: transaction committed but schema change aborted with error: (2BP01): could not remove enum value "chpw" as it is being used in a routine "udf_w1_328" (SQLSTATE 2BP01)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkloadMultiRegion
{
 "message": "Inspect Job Errors",
 "jobs": [
  {
   "jobId": 1167711504158654465,
   "status": "no errors reported"
  },
  {
   "jobId": 1167711460674404353,
   "status": "no errors reported"
  }
 ]
}
Schema Workload Stats
Total Schema Statements Executed = 133
Total Schema Statements Succeeded = 133
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 84
Total Transactions Rolled Back = 68
Total Transactions Executed = 152
    schema_change_external_test.go:136: backup, tracing data, and system table dumps in /artifacts/tmp/_tmp/69acf6b25649642ba7e134a867cc5b2c/logTestWorkloadMultiRegion491017237
    schema_change_external_test.go:145: running DROP with use_declarative_schema_changer = off
    panic.go:694: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/69acf6b25649642ba7e134a867cc5b2c/logTestWorkloadMultiRegion491017237
--- FAIL: TestWorkloadMultiRegion (116.55s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkloadMultiRegion.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkloadMultiRegion [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21317619?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21317619?buildTab=artifacts#/) on master @ [980e96a3482d2fed84c62da15ca95c0ee40c0639](https://github.com/cockroachdb/cockroach/commits/980e96a3482d2fed84c62da15ca95c0ee40c0639):


```
 }
}
    schema_change_external_test.go:183: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:183
        	            				pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:50
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED COMMIT ERROR; Received an unexpected commit error: ERROR: transaction committed but schema change aborted with error: (2BP01): could not remove enum value "fzcx" as it is being used in a routine "udf_w1_279" (SQLSTATE 2BP01)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:771
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.runSchemaChangeWorkload.runSchemaChangeWorkload.func5.func7
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:172
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_arm64.s:1447
        	            	Wraps: (4) ***UNEXPECTED COMMIT ERROR; Received an unexpected commit error
        	            	Wraps: (5) ERROR: transaction committed but schema change aborted with error: (2BP01): could not remove enum value "fzcx" as it is being used in a routine "udf_w1_279" (SQLSTATE 2BP01)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkloadMultiRegion
{
 "message": "Inspect Job Errors",
 "jobs": [
  {
   "jobId": 1167718737794957313,
   "status": "no errors reported"
  },
  {
   "jobId": 1167718551437377537,
   "status": "no errors reported"
  }
 ]
}
Schema Workload Stats
Total Schema Statements Executed = 113
Total Schema Statements Succeeded = 113
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 75
Total Transactions Rolled Back = 62
Total Transactions Executed = 137
    schema_change_external_test.go:136: backup, tracing data, and system table dumps in /artifacts/tmp/_tmp/69acf6b25649642ba7e134a867cc5b2c/logTestWorkloadMultiRegion26076776
    schema_change_external_test.go:145: running DROP with use_declarative_schema_changer = on
    panic.go:694: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/69acf6b25649642ba7e134a867cc5b2c/logTestWorkloadMultiRegion26076776
--- FAIL: TestWorkloadMultiRegion (78.71s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkloadMultiRegion.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/ccl/logictestccl/testdata/logic_test/multi_region
+++ b/pkg/ccl/logictestccl/testdata/logic_test/multi_region
@@ -1793,3 +1793,57 @@ statement ok
 DROP DATABASE "DelimitedDB_113782" CASCADE
 
 subtest end
+
+# Regression test #168543: creating a trigger on a REGIONAL BY TABLE table
+# or a GLOBAL table should succeed even when the trigger function references
+# the multi-region enum type. Previously, descriptor validation incorrectly 
+# rejected the region enum reference added by the trigger's type dependencies.
+subtest trigger_with_region_enum_reference
+
+statement ok
+CREATE DATABASE trigger_mr_db PRIMARY REGION "us-east-1" REGIONS "us-east-1", "ca-central-1", "ap-southeast-2";
+
+statement ok
+USE trigger_mr_db;
+
+statement ok
+CREATE TABLE rbt_primary (k INT PRIMARY KEY, v STRING);
+
+statement ok
+CREATE TABLE global_t (k INT PRIMARY KEY, v STRING) LOCALITY GLOBAL;
+
+statement ok
+CREATE OR REPLACE FUNCTION trigger_fn_mr()
+RETURNS TRIGGER LANGUAGE PLpgSQL AS $$
+DECLARE
+  r crdb_internal_region;
+BEGIN
+  r := 'us-east-1';
+  RETURN NEW;
+END;
+$$;
+
+# This should succeed — the trigger function references the region enum, but
+# that is a legitimate non-locality-config type dependency.
+statement ok
+CREATE TRIGGER tr_rbt BEFORE INSERT ON rbt_primary FOR EACH ROW EXECUTE FUNCTION trigger_fn_mr();
+
+statement ok
+CREATE TRIGGER tr_global BEFORE INSERT ON global_t FOR EACH ROW EXECUTE FUNCTION trigger_fn_mr();
+
+# Verify the triggers work.
+statement ok
+INSERT INTO rbt_primary VALUES (1, 'a');
+
+statement ok
+INSERT INTO global_t VALUES (1, 'a');
+
+statement ok
+USE system;
+
+statement ok
+DROP DATABASE trigger_mr_db CASCADE;
+
+subtest end
+
+

--- a/pkg/sql/catalog/multiregion/validate_table.go
+++ b/pkg/sql/catalog/multiregion/validate_table.go
@@ -117,12 +117,27 @@ func ValidateTableLocalityConfig(
 			break
 		}
 	}
-	columnTypesTypeIDs := catalog.MakeDescriptorIDSet(typeIDsReferencedByColumns...)
+	// Build a set of type IDs that are referenced by non-locality-config
+	// sources: columns, triggers, and policies. The region enum ID may appear
+	// in the full typeIDs set due to any of these, and all are legitimate.
+	nonLocalityTypeIDs := catalog.MakeDescriptorIDSet(typeIDsReferencedByColumns...)
+	for _, trigger := range desc.GetTriggers() {
+		for _, id := range trigger.DependsOnTypes {
+			nonLocalityTypeIDs.Add(id)
+		}
+	}
+	for _, policy := range desc.GetPolicies() {
+		for _, id := range policy.DependsOnTypes {
+			nonLocalityTypeIDs.Add(id)
+		}
+	}
 	switch lc := lc.Locality.(type) {
 	case *catpb.LocalityConfig_Global_:
 		if regionEnumIDReferenced {
-			// Omit views since they may reference the multi-region enum type of the base table.
-			if !columnTypesTypeIDs.Contains(regionsEnumID) && !desc.IsView() {
+			// The region enum may be referenced by a column type, trigger, or
+			// policy. We also omit views since they may reference the
+			// multi-region enum type of the base table.
+			if !nonLocalityTypeIDs.Contains(regionsEnumID) && !desc.IsView() {
 				return errors.AssertionFailedf(
 					"expected no region Enum ID to be referenced by a GLOBAL TABLE: %q"+
 						" but found: %d",
@@ -232,11 +247,11 @@ func ValidateTableLocalityConfig(
 			}
 		} else {
 			if regionEnumIDReferenced {
-				// It may be the case that the multi-region type descriptor is used
-				// as the type of the table column. Validations should only fail if
-				// that is not the case. We omit views since they may reference the
-				// multi-region enum type of the base table.
-				if !columnTypesTypeIDs.Contains(regionsEnumID) && !desc.IsView() {
+				// The region enum may be referenced by a column type, trigger, or
+				// policy on this table. Validations should only fail if none of
+				// those are the source of the reference. We also omit views since
+				// they may reference the multi-region enum type of the base table.
+				if !nonLocalityTypeIDs.Contains(regionsEnumID) && !desc.IsView() {
 					return errors.AssertionFailedf(
 						"expected no region Enum ID to be referenced by a REGIONAL BY TABLE: %q homed in the "+
 							"primary region, but found: %d",

